R/summarize_copilot.R

```diff
@@ -316,50 +326,49 @@ map_responses_to_cycles <- function(response_tbl,
         cycle.start_date,
         cycle.extended_start_date
       )
-    )
-
-  response_merged <- response_tbl %>%
+    ) %>%
+    # Select only the variables we want to bring into the response data; also,
+    # swap periods for underscores, otherwise the sql function won't work
+    select(cycle_id = cycle.uid,
+           cycle_ordinal = cycle.ordinal,
+           cycle_team_id = cycle.team_id,
+           cycle_extended_start_date = cycle.extended_start_date,
+           cycle_extended_end_date = cycle.extended_end_date)
+
+  # Prepare response data for merge
+  response_for_merge <- response_tbl %>%
+    # Add Triton classroom information
     left_join(triton.classroom, by = c(code = "classroom.code")) %>%
+    # Convert datetimes to dates
     mutate(
-      # Convert datetimes to dates; char > POSIXlt > char
-      # This will allow us to compare to cycle dates.
       created_date = created %>%
         strptime("%Y-%m-%d %H:%M:%S") %>%
-        strftime("%Y-%m-%d"),
-      cycle_id = NA, # Populated below.
-      cycle_ordinal = NA # Populated below.
-    )
+        strftime("%Y-%m-%d")
+    ) %>%
+    # Swap periods for underscores, otherwise the sql function won't work
+    rename(classroom_team_id = classroom.team_id)
+
+  # Merge
+  # By using sqldf, we can conduct a merge on dates in a fast way
+  # FYI, in SQL "between" is inclusive
+  response_merged <- sqldf(
+    "select * from response_for_merge
+    left join cycle_for_merge
+    on response_for_merge.classroom_team_id = cycle_for_merge.cycle_team_id
+    and response_for_merge.created_date between cycle_for_merge.cycle_extended_start_date and cycle_for_merge.cycle_extended_end_date"
```

Comment:
I thought for a bit "hey, shouldn't dplyr be able to do this?" but it really seems like it can't, because you're doing the logic in the join rule. So :+1: for finding the right tool for this job.

Reply:
Thanks!